crates/pathfinder/src/state/sync/pending.rs

```diff
@@ -28,96 +31,147 @@ pub async fn poll_pending(
+            .await
+            .context("Downloading pending block")
+    });
 
-        // Add a timeout to the pending state update query.
-        //
-        // This is work-around for the gateway constantly 503/502 on this query because
-        // it cannot calculate the state root on the fly quickly enough.
-        //
-        // Without this timeout, we can potentially sit here infinitely retrying this query internally.
-        let state_update = match tokio::time::timeout(
-            std::time::Duration::from_secs(3 * 60),
-            sequencer.state_update(BlockId::Pending),
-        )
-        .await
-        {
-            Ok(gateway_result) => gateway_result,
-            Err(_timeout) => {
-                tracing::debug!("Pending state update query timed out, exiting pending mode.");
-                return Ok((None, None));
+    loop {
+        let mut changed = false;
+
+        tokio::select! {
+            block = &mut block_task => {
+                match block.context("Joining block download task")?? {
+                    MaybePendingBlock::Block(block) if block.block_hash == head.0 => {
+                        // Sequencer `pending` may return the latest full block for quite some time, so ignore it.
+                        tracing::trace!(hash=%block.block_hash, "Found current head from pending mode");
+                    }
+                    MaybePendingBlock::Block(block) => {
+                        tracing::trace!(hash=%block.block_hash, "Found full block, exiting pending mode.");
+                        return Ok((Some(block), None));
+                    }
+                    MaybePendingBlock::Pending(pending) if pending.parent_hash != head.0 => {
+                        tracing::trace!(
+                            pending=%pending.parent_hash, head=%head.0,
+                            "Pending block's parent hash does not match head, exiting pending mode"
+                        );
+                        return Ok((None, None));
+                    }
+                    MaybePendingBlock::Pending(pending) => {
+                        let replace = match &prev_block {
+                            None => true,
+                            Some(prev) if pending.transactions.len() > prev.transactions.len() => true,
+                            _ => false,
+                        };
+
+                        if replace {
+                            prev_block = Some(Arc::new(pending));
+                            changed = true;
+                            tracing::trace!("Pending block data changed");
+                        } else {
+                            tracing::trace!("No change in pending block data");
+                        }
+                    },
+                }
+
+                let gateway_copy = sequencer.clone();
+                block_task = tokio::spawn(async move {
+                    tokio::time::sleep_until(Instant::now() + poll_interval).await;
```

Comment:
```suggestion
                    tokio::time::sleep(poll_interval).await;
```

Reply:
I replaced it with a proper interval instead. tokio provides a nice interval type but unfortunately one cannot clone it.